dids/src/main/kotlin/web5/sdk/dids/methods/dht/DidDht.kt

```diff
@@ -45,393 +79,500 @@ public class DidDhtConfiguration internal constructor(
-                val vmID = data["id"]!!
-                val keyBytes = Base64URL(data["k"]!!).decode()
-
-                // TODO(gabe): support other key types
-                val publicKeyJwk = when (data["t"]!!) {
-                  "0" -> Ed25519.bytesToPublicKey(keyBytes)
-                  "1" -> Secp256k1.bytesToPublicKey(keyBytes)
-                  else -> throw IllegalArgumentException("Unknown key type: ${data["t"]}")
-                }
+    message.addRecord(
+      TXTRecord(
+        Name("_did."), DClass.IN, ttl, rootRecordText.joinToString(";")
+      ), Section.ANSWER
+    )
 
-                verificationMethods += VerificationMethod.builder()
-                  .id(URI.create("$did#$vmID"))
-                  .type("JsonWebKey2020")
-                  .controller(URI.create(did))
-                  .publicKeyJwk(publicKeyJwk.toPublicJWK().toJSONObject())
-                  .build()
+    // if there are types, add a Resource Record for them
+    if (types != null) {
+      // convert types to integer values
+      val typeIndexes = types.map { it.index }
+      message.addRecord(
+        TXTRecord(
+          Name("_typ._did."), DClass.IN, ttl, "id=${typeIndexes.joinToString(",")}"
+        ), Section.ANSWER
+      )
+    }
 
-                keyLookup[name.split(".")[0].drop(1)] = "$did#$vmID"
-              }
-              // handle services
-              name.startsWith("_s") -> {
-                val data = parseTxtData(rr.strings.joinToString(","))
-                services += Service.builder()
-                  .id(URI.create("$did#${data["id"]!!}"))
-                  .type(data["t"]!!)
-                  .serviceEndpoint(data["uri"]!!)
-                  .build()
+    return message
+  }
+
+  /**
+   * Converts a DNS packet to a [DIDDocument] according to the did:dht spec
+   * https://tbd54566975.github.io/did-dht-method/#dids-as-a-dns-packet
+   *
+   * @param did The DID that the packet is for.
+   * @param msg The [Message] instance containing the DNS packet.
+   * @return A [Pair] containing the [DIDDocument] and a list of types.
+   * @throws IllegalArgumentException if the provided DID does not conform to the "did:dht" method.
+   */
+  @Suppress("CyclomaticComplexMethod")
+  public fun fromDnsPacket(did: String, msg: Message): Pair<DIDDocument, List<DidDhtTypeIndexing>> {
+    val doc = DIDDocument.builder().id(URI.create(did))
+
+    val verificationMethods = mutableListOf<VerificationMethod>()
+    val services = mutableListOf<Service>()
+    val types = mutableListOf<DidDhtTypeIndexing>()
+    val keyLookup = mutableMapOf<String, String>()
+
+    msg.getSection(Section.ANSWER).forEach { rr ->
+      when (rr) {
+        is TXTRecord -> {
+          val name = rr.name.toString()
+          when {
+            // handle verification methods
+            name.startsWith("_k") -> {
+              val data = parseTxtData(rr.strings.joinToString(""))
```

Comment:
Should there be some validation of the fields that `data` contains?

Reply:
that is what parseTxtData does...we don't enforce valid DID Documents here, just conversion to/from the packet form

if someone writes a junk packet that should be handled at a different layer